### DISCUSSION TITLE
Portenta H7 - SDRAM Configutration

### DISCUSSION THREAD

**Discussion by barisserbay** [micropython/micropython#12292] [General]
Hello all,

I have a [Portenta H7 Board](https://store.arduino.cc/products/portenta-h7) and I’ve installed the OpenMV micropython firmware via using the IDE. But I’m not using the camera, I only use the OpenMV micropython firmware couse it has all the data analysis libraries that I need on my project. (Ulab and tf)

I can use the flash memory withouth any problem but I think the firmware is not able to handle with the 8MB SDRAM on board. I've also tried with the firmware official micropython firmware. (Downloaded [here](https://micropython.org/download/ARDUINO_PORTENTA_H7/))

Here is what i see when I run the pyb.info() function:

```python
import pyb
pyb.info()
``` 
**Output:**
```
ID=24002500:09513033:36343638
DEVID=0x0450
REVID=0x2003
S=400000000
H=200000000
P1=100000000
P2=100000000
_etext=81fe538
_sidata=81fe540
_sdata=30000000
_edata=3000085c
_sbss=30004000
_ebss=300170c4
_sstack=20000000
_estack=20010000
_ram_start=30000000
_heap_start=300170c4
_heap_end=3003f0c4
_ram_end=30040000
qstr:
  n_pool=3
  n_qstr=69
  n_str_data_bytes=735
  n_total_bytes=1807
GC:
  160064 total
  11920 : 148144
  1=218 2=26 m=54
LFS free: 15575040 bytes
``` 
Also, I’ve use the gc library to understand how much RAM available on the board, please see the code and output below.
```python
print(f"Allocated Memory: {gc.mem_alloc()} Free Memory: {gc.mem_free()}")
``` 
**Output:**
_Allocated Memory: 5040 Free Memory: 155024_

Am I missing something or there is a problem with the firmware?

**Comment by jimmo** [comment]
@barisserbay The Portenta firmware in the main repo definitely enables SDRAM. (I can't speak for the OpenMV firmware).

It is also configured to do a sdram test on startup, importantly with "fail on error" set, so MicroPython wouldn't start up if there was a problem.

BUT... it doesn't look like it's configured to actually uses the SDRAM for the MicroPython heap.

Compare the Portenta config:
https://github.com/micropython/micropython/blob/master/ports/stm32/boards/ARDUINO_PORTENTA_H7/mpconfigboard.h#L229

to e.g. the F7 Discovery config, which overrides `MICROPY_HEAP_START/END` to use the sdram for the MicroPython heap:
https://github.com/micropython/micropython/blob/master/ports/stm32/boards/STM32F7DISC/mpconfigboard.h#L99

@iabdalkader Should we change this for the Portenta board? (and the Giga too?). Or is the intention with these boards to use the SRAM for heap and leave the SDRAM for other things?

**Reply by iabdalkader** [reply]
> Should we change this for the Portenta board? (and the Giga too?). Or is the intention with these boards to use the SRAM for heap and leave the SDRAM for other things?

In our firmware we use SRAM for MicroPython's heap, the SDRAM is used for frame buffers, intermediate allocation, TF buffers (arena), JPEG buffer etc.. there's not much use for a big MP heap, a fast one is more useful, but I think a feature that allows a secondary heap block was added a while ago we can look into it, add say 1 MB to MicroPython's heap (CC @kwagyeman might want to track this request)

For MicroPython you can use all of the SDRAM for heap if you want, we use separate linker scripts in the top dir (which gets input from different board config files) so changes here don't affect our firmware much.

**Comment by kwagyeman** [comment]
@barisserbay Please add it as a feature request here: https://github.com/openmv/openmv/issues

Adding 1MB for more MicroPython heap is possible. However, it's not the point of our system. But, if we give you the second heap block you could resize it then as you like to have more heap space yourself in a custom firmware.
